src/grammars/syntax.md

```diff
@@ -1,5 +1,453 @@
+
+Other **numbers of repetitions** can be indicated using curly brackets:
+
+```
+expr{n}           // exactly n repetitions
+expr{m, n}        // between m and n repetitions, inclusive
+
+expr{, n}         // at most n repetitions
+expr{m, }         // at least m repetitions
+```
+
+Thus `expr*` is equivalent to `expr{0, }`; `expr+` is equivalent to `expr{1,
+}`; and `expr?` is equivalent to `expr{0, 1}`.
+
+### Predicates
+
+Preceding an expression with an ampersand `&` or exclamation mark `!` turns it
+into a *predicate* that never consumes any input. You might know these
+operators as "lookahead" or "non-progressing".
+
+The **positive predicate**, written as an ampersand `&`, attempts to match its
+inner expression. If the inner expression succeeds, parsing continues, but at
+the *same position* as the predicate &mdash; `&foo ~ bar` is thus a kind of
+"AND" statement: "the input string must match `foo` AND `bar`". If the inner
+expression fails, the whole expression fails too.
+
+The **negative predicate**, written as an exclamation mark `!`, attempts to
+match its inner expression. If the inner expression *fails*, the predicate
+*succeeds* and parsing continues at the same position as the predicate. If the
+inner expression *succeeds*, the predicate *fails* &mdash; `!foo ~ bar` is thus
+a kind of "NOT" statement: "the input string must match `bar` but NOT `foo`".
+
+This leads to the common idiom meaning "any character but":
+
+```
+not_space_or_tab = {
+    !(                // if the following text is not
+        " "           //     a space
+        | "\t"        //     or a tab
+    )
+    ~ any             // then consume one character
+}
+
+triple_quoted_string = {
+    "'''"
+    ~ triple_quoted_character*
+    ~ "'''"
+}
+triple_quoted_character = {
+    !"'''"        // if the following text is not three apostrophes
+    ~ any         // then consume one character
+}
+```
+
+## Operator precedence and grouping (WIP)
+
+The repetition operators asterisk `*`, plus sign `+`, and question mark `?`
+apply to the immediately preceding expression.
+
+```
+"One " ~ "or " ~ "more. "+
+"One " ~ "or " ~ ("more. "+)
+    are equivalent and match
+"One or more. more. more. more. "
+```
+
+Larger expressions can be repeated by surrounding them with parentheses.
+
+```
+("One " ~ "or " ~ "more. ")+
+    matches
+"One or more. One or more. "
+```
+
+Repetition operators have the highest precedence, followed by predicate
+operators, the sequence operator, and finally ordered choice.
+
+```
+my_rule = {
+    "a"* ~ "b"?
+    | &"b"+ ~ "a"
+}
+    equivalent to
+my_rule = {
+      ( ("a"*) ~ ("b"?) )
+    | ( (&("b"+")) ~ "a" )
+}
+```
+
+## Start and end of input
+
+The rules `soi` and `eoi` match the *start* and *end* of the input string,
+respectively. Neither consumes any text. They only indicate whether the parser
+is currently at one edge of the input.
+
+For example, to ensure that a rule matches the entire input, where any syntax
```

Comment:
`soi` and `eoi` are also useful in order to get `whitespace` to match around the central rule. Might be worth mentioning, although it references a not-yet-introduced `whitespace`.